### PR TITLE
Optimize chocolatey when doing large numbers of operations, by batching and caching

### PR DESCRIPTION
**Pull request by jaymzjulian**
## Description

When we instrumented, chocolatey resources were a major (>30% run time) contibutor to our idempotent run time.  This reduces that to around 2%, via the following optimizations:

 * Cache choco list/search command output in order to reduce queries to the nuget servers.  This saves significant wall clock time when installing >50 packages

 * Attempt to batch queries as much as possible, either by walking the resource tree, or with a non-default resource property of "bulk_query".  This also significantly reduces load on upstream package servers, since queries are expensive.  Note that on _some_ nuget implementations, bulk_query gets a bad result from the server on v2.x - this is reproducable outside of chef, however.  The public repositories work for this, but it should always be non-default there since it's going to return a large list!

 * Grab package data from the disk rather than shelling out to chocolatey list, avoiding shell outs saving significant cycles on an idempotent runs

In our environment, this saves around 30% of the time of a chef run.

This is tested on chocolatey 1.2.1, 2.0.0, and 2.2.2

Note that this does NOT fix the existing rubocop issues - it doesn't add a pile of new ones, and it _does_ command and suppress the new one it introduces.  It might be worth another PR to generally clean this up for style...

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by jaymzjulian** [comment]
Boo, failing a test :(.  Will fix.

File: lib/chef/provider/package/chocolatey.rb [hunk 1]
```diff
@@ -162,7 +223,7 @@ def choco_exe
           @choco_exe ||= begin
               # if this check is in #define_resource_requirements, it won't get
               # run before choco.exe gets called from #load_current_resource.
-              exe_path = ::File.join(choco_install_path, "bin", "choco.exe")
+              exe_path = ::File.join(choco_install_path, 'bin', 'choco.exe')
```
Comment (jaymzh) [inline lib/chef/provider/package/chocolatey.rb:226]: That's FB internal style, Chef style is double quotes.

File: lib/chef/provider/package/chocolatey.rb [hunk 2]
```diff
@@ -214,6 +281,17 @@ def desired_name_versions
           Hash[*lowercase_names(new_resource.package_name).zip(desired_versions).flatten]
         end
 
+        def nested_package_resources(res)
+          package_collection = []
+          res.each do |child_res|
+            package_collection += nested_package_resources(child_res.resources)
+            next unless child_res.is_a? Chef::Resource::ChocolateyPackage
```
Comment (jaymzh) [inline lib/chef/provider/package/chocolatey.rb:288]:
```suggestion
            next unless child_res.is_a?(Chef::Resource::ChocolateyPackage)
```

File: lib/chef/provider/package/chocolatey.rb [hunk 3]
```diff
@@ -230,35 +308,96 @@ def cmd_args(include_source: true)
...
         def available_packages
-          return @available_packages if @available_packages
+          return @available_packages if !@available_packages.nil?
+          # @available_packages is per object - each resource is an object, meaning if you
+          # have a LOT of chocolatey package installs, then this quickly gets very slow.
+          # So we use @@choco_available_packages instead - BUT it's important to ensure that
+          # the cache is valid before you do this.  There are two cache items that can change:
+          # a) the sources - we check this with cache_is_valid?
+          if cache_is_valid? && @@choco_available_packages.is_a?(Hash) && 
+            @@choco_available_packages[new_resource.list_options]
```
Comment (jaymzh) [inline lib/chef/provider/package/chocolatey.rb:318]:
```suggestion
              @@choco_available_packages[new_resource.list_options]
```

File: lib/chef/provider/package/chocolatey.rb [hunk 4]
```diff
@@ -141,6 +154,39 @@ def query_command
...
+
+        # walk the collection to find peer resources to ensure that
+        # we get as much of the cache this run as possible.  Unified mode
+        # sub-resources will, of course, be incremental with this, since we can't
+        # grab them in advance, but even in that case we're still way, way
+        # more efficient with our queries...
+        #
+        # @return [Array] List of chocolatey packages referenced in the run list
+        def collect_package_requests(ignore_list: [])
+          return ['*'] if new_resource.bulk_query || ENV['BULK_QUERY_CHOCOLATEY'] == 'true'
```
Comment (jaymzh) [inline lib/chef/provider/package/chocolatey.rb:178]: The env var is an anti-pattern. You could make a Chef::Config option if you want something more global.

File: lib/chef/provider/package/chocolatey.rb [hunk 5]
```diff
@@ -230,35 +308,96 @@ def cmd_args(include_source: true)
...
+                Chef::Log.debug("Choco List Command: #{cmd}")
+
                 raw = parse_list_output(*cmd)
                 raw.keys.each_with_object({}) do |name, available|
                   available[name] = desired_name_versions[name] || raw[name]
                 end
-              end
-            @available_packages.merge! available_versions
+            end
+            @@choco_available_packages[new_resource.list_options].merge! available_versions
```
Comment (jaymzh) [inline lib/chef/provider/package/chocolatey.rb:367]:
```suggestion
            @@choco_available_packages[new_resource.list_options].merge!(available_versions)
```

**Review by jaymzh — CHANGES_REQUESTED** [review]
This is an interesting take.

It's not clear from the description, but we should note here, this is actually closer to how a lot of the other resources work, specifically yum/DNF.

I'm not a huge fan of the walking-the-resource collection, but it's not awful either.

Ultimately this works around a really big deficit in Chocolatey, best I can tell. What's kinda silly is that in yum/dnf land, we can just tell it download the latest metadata and then parse that.

I'm interested in @johnmccrae  and @mwrock 's opinion, but all in all I think this is a good call.

That said, @GeorgeWestwater - this is the kind of stuff I was talking about, the core engine is definitely not "done"... we need to be reworking a lot of these APIs. Every resource should have a "early hook" - if it exists in the resource collection we run the early hook before we converge the resources to do prep work once, instead of faking it on the first resource instance, which we do ALL over the place. I also wanted to re-work the notification system, and the way it hooks into the providers. THere's lots of good work to be done here.

**Comment by jaymzjulian** [comment]
Thanks for the feedback - appreciate it!  I'll get this stuff fixed up post haste

**Comment by jaymzjulian** [comment]
I updated this with suggestions, and also added an attribute to revert back to the previous behaviour for listing (`use_choco_list`), which defaults false but is there in case someone has a usecase where checking the on disk package repository doens't work for them for some reason.

I'm open to the idea that it should be `true` by default... I think the performance win is worthwhile having it on by default, but I'd absolutely be willing to hear disagreements and change that.

**Comment by tpowell-progress** [comment]
> I updated this with suggestions, and also added an attribute to revert back to the previous behaviour for listing (`use_choco_list`), which defaults false but is there in case someone has a usecase where checking the on disk package repository doens't work for them for some reason.
> 
> I'm open to the idea that it should be `true` by default... I think the performance win is worthwhile having it on by default, but I'd absolutely be willing to hear disagreements and change that.

My thinking on adding an option is that the default should generally not change behavior at all. However, in this case, this could also be viewed as a performance enhancement that you can turn off if you have an edge case that behaves badly with it on. If we choose `true` as a default, I would want to clearly call out mitigation procedures if your choco stuff goes south.

**Comment by GeorgeWestwater** [comment]
> That said, @GeorgeWestwater - this is the kind of stuff I was talking about, the core engine is definitely not "done"... 

Done and Feature Complete are not the same; I agree it is not "done". Software is never done

**Comment by tpowell-progress** [comment]
Tagged in @johnmccrae to take a peek as well.

**Comment by tpowell-progress** [comment]
@jaymzjulian still seeing a couple spellcheck errors. Looked like legit typos at a glance.

**Comment by jaymzjulian** [comment]
I believe the spelling errors in the comments are fixed now ;)

**Review by tpowell-progress — CHANGES_REQUESTED** [review]
@jaymzjulian I have just one question of clarification (on `lib/chef/provider/package/chocolatey.rb`)

File: lib/chef/provider/package/chocolatey.rb [hunk 6]
```diff
@@ -230,37 +309,124 @@ def cmd_args(include_source: true)
...
+        # vs calling choco list
+        #
+        # @return [Hash] name-to-version mapping of installed packages
+        def installed_packages_via_disk
+          @installed_packages ||= begin
+            targets = new_resource.name
+            target_dirs = []
+            # If we're using a single package name, have it at the head of the list
+            # so we can get more performance
+            target_dirs << targets.first.downcase if targets.length == 1
```
[the remark this answers is not in the document]
Comment (jaymzjulian) [reply lib/chef/provider/package/chocolatey.rb:412]: This is an optimization to try and push our particular package to the front of the queue if there is one and only one package specified.  If that doesn't happen, target_dirs instead gets filled at line 413 with:
`target_dirs += get_local_pkg_dirs(choco_lib_path)`

which gets all of the possible nupkg folders to search through.  This takes the 90% case to being "search half of the package folders on average" to "search one package on average"

**Review by tpowell-progress — APPROVED** [review]

**Review by tpowell-progress — APPROVED** [review]

**Review by tpowell-progress — APPROVED** [review]

**Pull request merged**